db/db_bolt.go

```diff
@@ -106,6 +106,32 @@ func (b *boltDB) Get(namespace string, key []byte) ([]byte, error) {
 	return nil, errors.Wrap(ErrIO, err.Error())
 }
 
+// Filter returns <k, v> pair in a bucket that meet the condition
+func (b *boltDB) Filter(namespace string, cond Condition) ([][]byte, [][]byte, error) {
+	var fk, fv [][]byte
+	if err := b.db.View(func(tx *bolt.Tx) error {
+		bucket := tx.Bucket([]byte(namespace))
+		if bucket == nil {
+			return errors.Wrapf(ErrNotExist, "bucket = %x doesn't exist", []byte(namespace))
```

Comment:
namespace not exist may be different from "ErrNotExist". It is possible that the bucket is empty or the entities in the bucket are all filtered by the "cond"

Reply:
don't quite get it, this is how we check if a bucket exist in the bolt DB ... 

even if the bucket is empty (no entries in it), returned "bucket" will not be nil if the name exist

Reply:
if we cannot return any entity from a bucket because of a filtering "cond", should we return ErrNotExist?

Reply:
i think this is not an error, but no entries match cond, in this case fk, fv will be nil, something like:

fk, fv, err := Filter()
if err != nil {
    return err
}

if len(fk) ==0 or len(fv) == 0 {
   // no match, do nothing
   return nil
}

process(fk, fv)
return nil